### PR TITLE
Deprecate the Microsoft.AspNetCore.App metapackage in favor of targeting and runtime packs

### LINKED ISSUE #6501: Create targeting pack for ASP.NET Core
For .NET Core 3.0, we expect to separate targeting packs, which contain the assets needed at build time when targeting .NET Core 3.0 (for example, reference assemblies and intellisense documentation) from "runtime packs", which will include the platform-specific runtime components for .NET Core 3.  See dotnet/cli#10006, and [dotnet/designs#50](https://github.com/dotnet/designs/pull/50/files?short_path=ebf265d#diff-ebf265d0476041dd8cab82fa24ab5b0e)

We will have targeting pack for each shared framework (.NET Core, ASP.NET Core, and WindowsDesktop), as well as for .NET Standard.  This issue tracks the work required for the ASP.NET Core shared framework.

The work required includes:

- A new NuGet package including the targeting pack assets.  This should include the reference assemblies, intellisense documentation, and platform manifest file.  The current straw man proposal for the package ID is `Microsoft.NETCore.App.Ref`
- Installers for each supported installer format (msi, deb, pkg, and rpm), which lay out the targeting pack files under `<DOTNET_ROOT>/packs/<PackageId>/version`.  These installers will be bundled or depended on by the corresponding .NET Core SDK installers.

Some of the details (such as package layout, names, etc) are still [under design](https://github.com/dotnet/designs/pull/50/files?short_path=ebf265d#diff-ebf265d0476041dd8cab82fa24ab5b0e) and may change slightly.

### PR DESCRIPTION
**Pull request by natemcmaster**
Resolves #6501 

This is all about how we build and redistribute aspnetcore. It doesn't change the way users reference aspnetcore.

Changes
* Stop producing Microsoft.AspNetCore.App, and runtime.$(rid).Microsoft.AspNetCore.App
* Generate the shared framework without using 'NuGet' restore
* Put the platform manifest into the targeting pack (data/Microsoft.AspNetCore.App.PlatformManifest.txt)
* Removes the need to generate packages so we can figure out what belongs in the shared framework
* Create well-known, shared intermediate directories that installers can use to bundle content
 
cc @nguerrera @dagood @dsplaisted

### CODE REVIEW

**Comment by dsplaisted** [comment]
We need to do https://github.com/dotnet/cli/issues/10442 before you stop producing the runtime.$(rid).Microsoft.AspNetCore.App packages

**Comment by natemcmaster** [comment]
I see https://github.com/dotnet/cli/issues/10442 is marked as `<blocked>`. Do you have a rough ETA?

**Comment by nguerrera** [comment]
I think the `blocked` was because we didn't have the new packages so that is now unblocked. Also, we don't need the file list to unblock this because we kept nuget folder conventions.

**Comment by natemcmaster** [comment]
> we don't need the file list

Which file list, the platform manifest?

File: Directory.Build.targets
Comment (dougbu) [inline Directory.Build.targets]: Should be `OR`

File: docs/ProjectProperties.md
Comment (dougbu) [inline docs/ProjectProperties.md]: Property needs a new name if you really mean "and is not &hellip;". Isn't "not available as a NuGet package" true for all shared framework components these days?

IOW property name makes sense only so long it means the property ships in the shared framework (no caveats).

**Review by dougbu — APPROVED** [review]
One bit is confusing me but, otherwise (once unblocked and builds pass), :shipit:

File: Directory.Build.targets
Comment (natemcmaster) [reply Directory.Build.targets]: `AND` is deliberate. This ensures we don't produce packages for shared-framework-only assemblies like Mvc.Core, but still do for shared-framework-and-package assemblies like SignalR

File: docs/ProjectProperties.md
Comment (natemcmaster) [reply docs/ProjectProperties.md]: > Isn't "not available as a NuGet package" true for all shared framework components these days?

Nope. Here are the exceptions: https://github.com/aspnet/AspNetCore/blob/f19e8da1daf2808f3bbb870141304b73e76a547a/eng/SharedFramework.Local.props#L10-L17

**Review by pakrym — APPROVED** [review]

File: Directory.Build.targets
Comment (natemcmaster) [reply Directory.Build.targets]: Decide to back this out. After closer inspection of the nupkg's produced by the build, we need to keep producing the intermediate nupkg's for now. We can tackle this again when we do #4257

**Comment by pakrym** [comment]
cc @anurse

**Comment by natemcmaster** [comment]
After this merges, anything still using a package reference to Microsoft.AspNetCore.App will stop getting updates. The new packages, Microsoft.AspNetCore.App.Ref and Microsoft.AspNetCore.App.Runtime.$(rid) should not be referenced directly...they're not mean to be used as a `<PackageReference>`. Instead, consumers should use `<FrameworkReference>`.

@nguerrera @dsplaisted - if we merge this to aspnet today, will you be able to make changes to react to this in dotnet/core-sdk? I took a look, and it seems the repo is still using Microsoft.AspNetCore.App to flow dependency updates and find the list of RIDs the aspnetcore runtime supports. 

cc @sebastienros - benchmarking infrastructure

**Comment by davidfowl** [comment]
Are the sdk changes  already in to consume these changes?

**Comment by natemcmaster** [comment]
No, that's why I made the previous comment. Not sure if we're expected to
do this or if the SDK team is already on it.

>

**Comment by davidfowl** [comment]
Ah then we shouldn’t stop producing the old shared framework until that’s the case.

**Comment by dsplaisted** [comment]
> Not sure if we're expected to do this or if the SDK team is already on it.

We will be updating the SDK to consume the updated packages

**Comment by natemcmaster** [comment]
I'd like to merge this before master diverges too much from this PR. Got tired of waiting for the SDK, so here you go: https://github.com/dotnet/core-sdk/pull/754

**Review by sebastienros — APPROVED** [review]

**Comment by davidfowl** [comment]
@natemcmaster This change is problematic. Do libraries authors build packages that use ASP.NET Core during the 3.0 time frame? This will affect our GRPC development experience if the ref pack can't be used as a substitute.

**Comment by natemcmaster** [comment]
Quoting myself from above

> This is all about how we build and redistribute aspnetcore. It doesn't change the way users reference aspnetcore.

Keep using `<FrameworkReference>`

**Comment by dsplaisted** [comment]
@davidfowl Libraries should not have a package dependency on Microsoft.AspNetCore.App.  We need https://github.com/NuGet/Home/issues/7342 and https://github.com/dotnet/cli/issues/10666 for packages to be able to declare dependencies on the AspNetCore shared framework.  Until then, apps that reference libraries that depend on ASP.NET Core will need to have a direct FrameworkReference to Microsoft.AspNetCore.App.

**Comment by davidfowl** [comment]
I see, libraries just use the FrameworkReference for compile assets and those won't persist. OK. I see that's already the case for GRPC https://www.nuget.org/packages/Grpc.AspNetCore.Server/0.1.19-pre1.

**Comment by nguerrera** [comment]
Aside: GRPC's dependency on Microsoft.NETCore.Platforms is a bug (fixed in preview3) and can lead to downgrades.

**Pull request merged**
